feat: NFR Tests Part 1 (#171)
lilweege/SyntaxSentinels: docs/VnVPlan/VnVPlan.tex

  \end{enumerate}

  \item{test-UH-3}
  
  NFR: UH-PI2

  Type: Manual
            
  Initial State: Application on home page, ready for use
            
  Input: Tester engagement
            
  Output: All identified instances of foreign languages used
            
  How test will be performed: 
  \begin{enumerate}
    \item The tester will look through each page, and test functionality
    \item They will verify that all pages and prompts are in the correct language (English (US))
  \end{enumerate}
  
  \end{enumerate}

\subsubsection{Versioning}
These requirements are for versioning, in order to roll back to old iterations
if needed. These will mainly be tested by checking if old versions are saved.
Note that this check will be done after each iteration of the model

\paragraph{Validate old versions of project}
\begin{enumerate}
  \item{test-V-1}

  NFR: MS-M1, OR-R2
  
  Type: Manual
            
  Initial State: N/A
            
  Input/Condition: Tester engagement
            
  Output/Result: The way to find old iterations/versions of project
            
  How test will be performed: 
  \begin{enumerate}

>     \item A tester must check if old iterations are stored properly (on Github, for example)

Reviewer: Old iteration of what? We should store old models with appropriate semantic versioning.

\subsubsection{Privacy/Security}
These requirements ensure that the system is secure, and does not violate any
privacy rights. The tests for these requirements will revolve around doing
audits to ensure system integrity.

\begin{description}
  \item[Data Retention Audit:] We will manually check the file system and any
  logs to look for any user-generated data, or personally identifiable data.
  This audit may also be done in a test environment to check live logs of the
  system.
  \item[Endpoint Security Check:] A tester will invoke backend endpoints and
  ensure that they are secure. They must also check that the website uses https
  encryption (padlock symbol in address bar)
  \item[Legal Audit:] We will manually check the system to ensure that it
  adheres to legal policies defined in the SRS, such as PIPEDA and FIPPA. This
  audit will be done regularly, and with each new version release, to the best
  of our abilities.
\end{description}

\paragraph{Validate Zero Data Retention and data encryption}
\begin{enumerate}
  \item{test-PS-1}
  
  NFR: SR-P1

  Type: Manual, Data Retention Audit

  Initial State: Application ready for use

  Input/Condition: Tester engagement

>   Output/Result: List of all identified components that have data stored 

Reviewer: ```suggestion
  Output/Result: A list of all identified user-generated or personally identifiable data
```

  How test will be performed:
  \begin{enumerate}
    \item A tester will enable logging in the components of the application

>     \item They will do a run through of the different functionalities

Reviewer: ```suggestion
    \item They will do a run through of the different functionalities, including making a submission and analyzing a generated report file
```

>     \item With each function, the tester will check logs and file systems and

Reviewer: ```suggestion
    \item With each function, the tester will check logs and filesystem, and then
```

>     search for any type of data being stored

Reviewer: ```suggestion
    search for any user-generated or personally identifiable data
```

  \end{enumerate}

  \item{NFR-PS2}
  
  NFR: SR-P2 

  Type: Manual 

  Initial State: Application ready for use

  Input/Condition: Backend endpoint invocation

>   Output/Result: Identified components that do not adhere to security requirements

Reviewer: ```suggestion
  Output/Result: A list of exposed endpoints that do not support in-transit encryption
```

  How test will be performed:
  \begin{enumerate}
    \item A tester will invoke backend end points using HTTPS protocol
    \item They will ensure that the response provides relevant security headers
  \end{enumerate}

  \item{test-PS-3}
  
  NFR: CR-L1, CR-L2, CR-L3, CR-S2, CR-S5

  Type: Manual, Legal Audit

  Initial State: Application ready for use

  Input/Condition: Tester engagement
  

>   Output/Result: List of vulnerabilites and breach of regulations

Reviewer: ```suggestion
  Output/Result: List of components of the system that breach any regulations, with an accompanying explanation of the how the breach could manifest itself
```

  How test will be performed:
  \begin{enumerate}

>     \item A tester will inspect the code repository and system application against laws and regulations

Reviewer: ```suggestion
    \item A third party tester will inspect the code repository and system application against laws and regulations, noting any component of the system that may potentially infringe upon any data or privacy laws, intellectual property, or academic integrity
```

>     \item All discrepancies between the law and the system will be identified, and updated promptly.

Reviewer: This isn't part of the test
```suggestion
```

\subsubsection{Development Standards}
These requirements ensure that the system is built in a way that adheres to
standard software development principles. This ensures that the code is
maintainble, scalable, and reliable.

\begin{description}
  \item[Code Inspection:] The code will be inspected to make sure the system is
  built adhering to software development standards.
  \item[Respository Inspection:] The Github repository for this project will be
  checked to make sure bugs are listed as issues to keep track of what needs to
  be done in the system 
\end{description}

\paragraph{Validate Development Standards}
\begin{enumerate}
  \item{test-DS-1}
  
  NFR: MS-M3

  Type: Manual, Repository Inspection

  Initial State: Development temporarily paused

  Output/Result: Indication of bugs being tracked on Github

  How test will be performed:
  \begin{enumerate}
    \item A tester will look at the 'issues' section of the github repository 
    \item They will check that there are bugs that exist as issues
    \item If there aren't, they will consult the team and ask if developers are
    issuing bugs or if there are simply no bugs
  \end{enumerate}

  \item{test-DS-2}
  

>   NFR: CR-SC1

Reviewer: ```suggestion
  NFR: MS-S1, CR-SC1
```

  Type: Manual, Code Inspection

  Initial State: Development temporarily paused

  Output/Result: Indication of components not adhering to software development standards

  How test will be performed:
  \begin{enumerate}

>     \item Team members will regularly inspect the code and ensure all components follow software development standards

Reviewer: ```suggestion
    \item Team members will regularly inspect the code and ensure all components follow software development standards, including SOLID principles and the existence of documentation and tests
```

    \item If there are components that don't they will consult the other team members to discuss issues with that component.
  \end{enumerate}
\end{enumerate}

\subsubsection{Documentation}
This section is to ensure that documentation exists to guide users and help them troubleshoot issues they are having.

\begin{description}
  \item[Documentation Review:] A tester will look through the webpage, and
  ensure that documentation exists. Then, they must check to make sure the
  documentation is accurate, and not misleading. 
\end{description}

\paragraph{Validate Program, help, and training documentation exists and is accurate}
\begin{enumerate}
  \item{test-D-1}
  

>   NFR: MS-S1, UH-L2, OE-P2

Reviewer: I think MS-S1 should be included in the previous test, it is more so about internal API documentation than user documentation.
```suggestion
  NFR: UH-L2, OE-P2
```

Reviewer: @MohammadMohsinKhan FYI

  Type: Manual, Documentation Review

  Initial State: Application ready for use

  Output/Result: Certification that documentation is valid

  How test will be performed:
  \begin{enumerate}

>     \item A developer will review the documentation on the application. They must first ensure that documentation exists, and is easy to find
>     \item They must then verify that the documentation is correct, and complete
>     \item Anything missing or incorrect about the documentation will be brought up to the team immediately.

Reviewer: ```suggestion
    \item A developer will navigate to the user documentation and help pages within the application. They must first ensure that documentation exists, and is easy to find
    \item They must then verify that the documentation is correct, complete, uses clear non-technical language, and is helpful for users
```

    \item A list of commonly used functions will be created (System is not large, therefore this can be done manually)
    \item The user will navigate to each of those functions, and note how many clicks were required to get there
  \end{enumerate}

  \item{test-UH-3}
  
  NFR: UH-PI2

  Type: Manual
            
  Initial State: Application on home page, ready for use
            
  Input: Tester engagement
            
  Output: All identified instances of foreign languages used
            
  How test will be performed: 
  \begin{enumerate}
    \item The tester will look through each page, and test functionality
    \item They will verify that all pages and prompts are in the correct language (English (US))
  \end{enumerate}
  
  \end{enumerate}

\subsubsection{Versioning}
These requirements are for versioning, in order to roll back to old iterations
if needed. These will mainly be tested by checking if old versions are saved.
Note that this check will be done after each iteration of the model

\paragraph{Validate old versions of project}
\begin{enumerate}
  \item{test-V-1}

  NFR: MS-M1, OR-R2
  
  Type: Manual
            
  Initial State: N/A
            

>   Input/Condition: Tester engagement

Reviewer: What does this mean lol

Author: The tester will be doing things with the app to test it manually

be involved? Be specific.}

\subsubsection{Look and Feel Requirements}
These tests are to ensure that the way the front end is designed adheres to the
defined requirements with respect to layout, typography, component alignment,
etc. 

\begin{description}
  \item[Front End Component Review:] We will engage a developer to look through
  the code and interface and ensure components are implemented as they are
  defined in the SRS.
  \item[Accessibility Check:] A developer will review if accessibility tools

>   exist or not within the application. Important tools are screen readers and

Reviewer 2: mention aria-label for screen reader

Reviewer 2: also do we need to specify who the developer is? just wondering how specific we should be

Author: No, just need that _a_ developer is in charge of it

  alt text for images. Furthermore, the developer must check for sufficient
  colour contrast between components.
\end{description}

> 

Reviewer 3: fixed

\begin{enumerate}

\item{NFR-UI1}

> NFR: LF-AR1, LF-AR2, LF-AR3, LF-SR1, LF-SR2, LF-SR3, LF-SR4

Reviewer 2: why are all of these grouped? this feels way too vague @lucasichen thoughts?

Type: Manual, Front End Review
					
Initial State: Application on home page, ready for use
					
Input/Condition: Tester engagement
					
Output/Result: List of all identified inconsistencies that are found that do not adhere to SRS defined requirements.
					
How test will be performed: 
\begin{enumerate}
  \item A developer will review the code to make sure the look and feels requirements were implemented
  \item They will then look at the application as well to ensure that the implementation was made correctly
\end{enumerate}
					
\item{NFR-UI2}

NFR: CR-SC3

Type: Manual, Accessibility Check
					
Initial State: Application on home page, ready for use
					
Input: Tester engagement
					
Output: List of missing accessibility considerations that have been overlooked
					
How test will be performed: 

> \begin{enumerate}

Reviewer 2: again, same comment as above, this test is just way too vague. Mention what enables screen readers (aria-labels mainly). Mention how the "developer" should make sure alt text exists for every image

  \item{NFR-UH1}
  
  NFR: UH-E1, UH-E3, UH-L1, UH-UP1, UH-UP2, OE-P1
  
  Type: Manual, User Test Demo
            
  Initial State: Application on home page, ready for use
            
  Input/Condition: User engagement
            
  Output/Result: Subjective feedback for how well application scores on different requirements relating to user experience
            
  How test will be performed: 
  \begin{enumerate}
    \item A user will use the application, going through the onboarding process
    and doing a demo of the application
    \item After a full demo, the user will be asked to provide feedback on
    different aspects of usability, such as ease of use, app clarity, and prompt
    politeness.
  \end{enumerate}
            
  \item{NFR-UH2}
  
  NFR: UH-E2
  
  Type: Manual
            
  Initial State: Application on home page, ready for use
            
  Input: Tester engagement
            
  Output: A count for the number of clicks required to navigate to different functionality
            
  How test will be performed: 
  \begin{enumerate}
    \item A list of commonly used functions will be created (System is not large, therefore this can be done manually)

>     \item The developer will navigate to each of those functions, and note how many clicks were required to get there

Reviewer 2: developer or user?

  \end{enumerate}

  \item{NFR-UH3}
  
  NFR: UH-PI2

  Type: Manual
            
  Initial State: Application on home page, ready for use
            
  Input: Tester engagement
            
  Output: All identified instances of foreign languages used
            
  How test will be performed: 
  \begin{enumerate}
    \item The tester will look through each page, and test functionality

>     \item They will verify that all pages and prompts are in the correct language

Reviewer 2: maybe mention the correct language is English (US)

be involved? Be specific.}

\subsubsection{Look and Feel Requirements}
These tests are to ensure that the way the front end is designed adheres to the
defined requirements with respect to layout, typography, component alignment,
etc. 

\begin{description}
  \item[Front End Component Review:] We will engage a developer to look through
  the code and interface and ensure components are implemented as they are
  defined in the SRS.
  \item[Accessibility Check:] A developer will review if accessibility tools
  exist or not within the application. Important tools are screen readers and
  alt text for images. Furthermore, the developer must check for sufficient
  colour contrast between components.
\end{description}

\begin{enumerate}

\item{NFR-UI1}

NFR: LF-AR1, LF-AR2, LF-AR3, LF-SR1, LF-SR2, LF-SR3, LF-SR4

Type: Manual, Front End Review
					
Initial State: Application on home page, ready for use
					
Input/Condition: Tester engagement
					

> Output/Result: List of all identified inconsistencies that are found that do not adhere to SRS defined requirements.

Reviewer: Can we list them explicitly? I.e. For each NFR we should mention what we should be adhering to.

Reviewer: Addressed in 91e5c568de4fc8efc5f61fffd99b982b9db10ebc

					
How test will be performed: 
\begin{enumerate}
  \item A developer will review the code to make sure the look and feels requirements were implemented

>   \item They will then look at the application as well to ensure that the implementation was made correctly

Reviewer: ```suggestion
  \item They will then check that the application to ensure that it adheres to all look and feel requirements
```

\end{enumerate}
					
\item{NFR-UI2}

NFR: CR-SC3

Type: Manual, Accessibility Check
					
Initial State: Application on home page, ready for use
					
Input: Tester engagement
					
Output: List of missing accessibility considerations that have been overlooked
					
How test will be performed: 
\begin{enumerate}
  \item The developer will review the code and application to ensure that accessibility considerations have been implemented

>   \item They will verify that all images have alt text

Reviewer: ```suggestion
  \item They will verify that all images have alt text
  \item They will verify that all text has sufficient contrast and sizing using Google Chrome devtools
```

These requirements are intended to build a foundation for the user's experience
on the application. Thus, the only real way to get feedback the user experience
is to have someone use the application and provide feedback.

\begin{description}
  \item[User Test Demo:] We will engage a user to test the application, bringing
  them through a new user process, and ask for feedback on specific areas
  relevant to the requirements that were set.
\end{description}

\begin{enumerate}

  \item{NFR-UH1}
  
  NFR: UH-E1, UH-E3, UH-L1, UH-UP1, UH-UP2, OE-P1
  
  Type: Manual, User Test Demo
            
  Initial State: Application on home page, ready for use
            
  Input/Condition: User engagement
            
  Output/Result: Subjective feedback for how well application scores on different requirements relating to user experience
            
  How test will be performed: 
  \begin{enumerate}
    \item A user will use the application, going through the onboarding process
    and doing a demo of the application
    \item After a full demo, the user will be asked to provide feedback on
    different aspects of usability, such as ease of use, app clarity, and prompt

>     politeness.

Reviewer: Can we specify what they're being asked? We should put the exact wording of the question such as "On a scale of 1 to 10, how would you rate X". Then the output/result section can be "User feedback scores in categories X, Y, Z on a scale of 1 to 10"

Reviewer: Include one question for each requirement that this test is for

Author: This belongs in a survey we have yet to make

Reviewer: It needs to be *specific*. This document is meant to include that survey if there is one

Reviewer: ![{39F50886-B2B5-4A13-963B-719318B4FA75}](https://github.com/user-attachments/assets/71ff3f3f-3b61-4363-93ac-c3f5e485db5d)

  \end{enumerate}
            
  \item{NFR-UH2}
  
  NFR: UH-E2
  
  Type: Manual
            
  Initial State: Application on home page, ready for use
            
  Input: Tester engagement
            
  Output: A count for the number of clicks required to navigate to different functionality
            
  How test will be performed: 
  \begin{enumerate}

>     \item A list of commonly used functions will be created (System is not large, therefore this can be done manually)

Reviewer: What are the commonly used functions? We need to list them out.
Here's what I had in mind (only really these two matter):
- Make a submission
- View a generated report

Author: I literally said a list of commonly used functions will be created

Reviewer: You can't just say the list will be created, we have to actually have the list. That's the point of the test.

Reviewer: Addressed in 917cc17c45140145e766d64cb16526099f24393d

Author: As in, the list will be made when we do the actual tests

    \item The user will navigate to each of those functions, and note how many clicks were required to get there
  \end{enumerate}

  \item{NFR-UH3}
  
  NFR: UH-PI2

  Type: Manual
            
  Initial State: Application on home page, ready for use
            
  Input: Tester engagement
            
  Output: All identified instances of foreign languages used
            
  How test will be performed: 
  \begin{enumerate}

>     \item The tester will look through each page, and test functionality

Reviewer: They don't need to test functionality, just check that it's in English

Author: How will they seek every prompt without going through functionality?

Reviewer: They don't have to necessarily perform the function, just go to the page

Reviewer: Addressed in 5f4b6915d17ae7f03a474df7db5470520e5bac19
